src/fastalloc/mod.rs

```diff
@@ -0,0 +1,1854 @@
+                    let slot = if self.vreg_spillslots[vreg.vreg()].is_valid() {
+                        self.vreg_spillslots[vreg.vreg()]
+                    } else {
+                        self.vreg_spillslots[vreg.vreg()] = self.allocstack(&vreg);
+                        self.vreg_spillslots[vreg.vreg()]
+                    };
+                    let slot_alloc = Allocation::stack(slot);
+                    self.add_move_later(
+                        inst,
+                        preg_alloc,
+                        slot_alloc,
+                        vreg.class(),
+                        InstPosition::Before,
+                        true
+                    );
+                    self.add_move_later(
+                        inst,
+                        slot_alloc,
+                        preg_alloc,
+                        vreg.class(),
+                        InstPosition::After,
+                        false,
+                    );
+                }
+            }
+        }
+        trace!("Done adding edits for clobbered registers");
+    }
+
+    /// If instruction `inst` is a branch in `block`,
+    /// this function places branch arguments in the spillslots
+    /// expected by the destination blocks.
+    /// 
+    /// The process used to do this is as follows:
+    /// 
+    /// 1. Move all branch arguments into corresponding temporary spillslots.
+    /// 2. Move values from the temporary spillslots to corresponding block param spillslots.
+    /// 
+    /// These temporaries are used because the moves have to be parallel in the case where
+    /// a block parameter of the successor block is a branch argument.
+    fn process_branch(&mut self, block: Block, inst: Inst) {
+        
+        // Used to know which temporary spillslot should be used next.
+        let mut next_temp_idx = PartedByRegClass { items: [0, 0, 0] };
+
+        fn reset_temp_idx(next_temp_idx: &mut PartedByRegClass<usize>) {
+            next_temp_idx[RegClass::Int] = 0;
+            next_temp_idx[RegClass::Float] = 0;
+            next_temp_idx[RegClass::Vector] = 0;
+        }
+
+        // In the case where the block param of a successor is also a branch arg,
+        // the reading of all the block params must be done before the writing.
+        // This is necessary to prevent overwriting the branch arg's value before
+        // placing it in the corresponding branch param spillslot.
+
+        for succ in self.func.block_succs(block).iter() {
+            self.alloc_slots_for_block_params(*succ);
+        }
+
+        for (succ_idx, _) in self.func.block_succs(block).iter().enumerate() {
+            // Move from branch args spillslots to temporaries.
+            //
+            // Consider a scenario: block X branches to block Y and block Y branches to block X.
+            // Block Y has block param vp and block X uses virtual register va as the branch arg for vp.
+            // Block X has an instruction that uses vp.
```

Comment:
It's impossible for Block X to use vp because it isn't dominated by the definition of vp which is in Block Y. SSA requires that there only be one def per vreg and that the def dominates all uses. A block argument counts as a def of that vreg.

Reply:
Whoops, I think I was unclear there. I meant that Y dominates X and X has a loop back to Y. Something like:

```
block entry
  goto Y(defines some value for vp)

block Y(vp)
  goto X

block X
  goto Y(va)
```
So we have: entry -> Y -> X then back to -> Y